dataplane/forwarding/protocol/mpls/mpls.go

```diff
@@ -0,0 +1,212 @@
+	desc   *protocol.Desc
+	labels []*mplsLabel
+}
+
+func (m *mpls) Header() []byte {
+	b := make([][]byte, len(m.labels))
+	for p, h := range m.labels {
+		b[p] = h.hdr
+	}
+	return bytes.Join(b, []byte{})
+}
+
+func (m *mpls) Trailer() []byte {
+	return nil
+}
+
+func (m *mpls) ID(instance int) fwdpb.PacketHeaderId {
+	return fwdpb.PacketHeaderId_PACKET_HEADER_ID_MPLS
+}
+
+const (
+	labelOffset    = 0
+	labelByteLen   = 4  // To make dealing with label easier, encode it as uint32.
+	labelBitOffset = 12 // The label is the leftmost 20 bits of the header, so shift 12 bits.
+	labelBitLen    = 20
+	tcOffset       = 2
+	tcByteLen      = 1
+	tcBitOffset    = 1
+	tcBitLen       = 3
+	ttlOffset      = 3
+	ttlByteLen     = 1
+	botOffset      = 2
+	botByteLen     = 1
+	botBitOffset   = 0
+	botBitLen      = 1
+)
+
+func (m *mpls) field(id fwdpacket.FieldID) frame.Field {
+	l := m.labels[id.Instance]
+	if id.IsUDF {
+		return protocol.UDF(l.hdr, id)
+	}
+	switch id.Num {
+	case fwdpb.PacketFieldNum_PACKET_FIELD_NUM_MPLS_LABEL:
+		return l.hdr.Field(labelOffset, labelByteLen)
+	case fwdpb.PacketFieldNum_PACKET_FIELD_NUM_MPLS_TC:
+		return l.hdr.Field(tcOffset, tcByteLen)
+	case fwdpb.PacketFieldNum_PACKET_FIELD_NUM_MPLS_TTL:
+		return l.hdr.Field(ttlOffset, ttlByteLen)
+	default:
+		return nil
+	}
+}
+
+func (m *mpls) Field(id fwdpacket.FieldID) ([]byte, error) {
+	field := m.field(id)
+	if field == nil {
+		return nil, fmt.Errorf("ethernet: Field failed, field %v does not exist", id)
+	}
+	if id.IsUDF {
+		return field.Copy(), nil
+	}
+	switch id.Num {
+	case fwdpb.PacketFieldNum_PACKET_FIELD_NUM_MPLS_LABEL:
+		return field.BitField(labelBitOffset, labelBitLen), nil
+	case fwdpb.PacketFieldNum_PACKET_FIELD_NUM_MPLS_TC:
+		return field.BitField(tcBitOffset, tcBitLen), nil
+	default:
+		return field.Copy(), nil
+	}
+}
+
+func (m *mpls) UpdateField(id fwdpacket.FieldID, op int, arg []byte) (bool, error) {
+	field := m.field(id)
+	if field == nil {
+		return false, fmt.Errorf("mpls: UpdateField failed, field %v does not exist", id)
+	}
+	if op == fwdpacket.OpDec && id.Num == fwdpb.PacketFieldNum_PACKET_FIELD_NUM_MPLS_TTL {
+		pttl := uint8(field.Value())
+		nttl := pttl - 1
+		field.SetValue(uint(nttl))
+		return false, nil
+	}
+	if op != fwdpacket.OpSet {
+		return false, fmt.Errorf("only SET is supported")
+	}
+
+	if id.IsUDF && op == fwdpacket.OpSet {
+		return true, field.Set(arg)
+	}
+	switch id.Num {
+	case fwdpb.PacketFieldNum_PACKET_FIELD_NUM_MPLS_LABEL:
+		field.SetBits(labelBitOffset, labelBitLen, uint64(binary.BigEndian.Uint32(arg)))
+		return true, nil
+	case fwdpb.PacketFieldNum_PACKET_FIELD_NUM_MPLS_TC:
+		field.SetBits(tcBitOffset, tcBitLen, uint64(arg[0]))
+		return true, nil
+	default:
+		return true, field.Set(arg)
+	}
+}
+
+func (m *mpls) Remove(id fwdpb.PacketHeaderId) error {
+	if len(m.labels) > 1 {
+		m.labels = m.labels[1:]
+		return nil
+	}
+	m.labels = nil
+	return nil
+}
+
+func (m *mpls) Modify(id fwdpb.PacketHeaderId) error {
```

Comment:
id is not referred in the method. What is modified?

Reply:
modify adds a label to the stack. the id is always mpls